xapian-letor/tests/api_letor.cc

```diff
@@ -98,6 +102,60 @@ DEFINE_TESTCASE(createfeaturevector, generated)
     return true;
 }
 
+DEFINE_TESTCASE(preparetrainingfile, generated)
+{
+    string db_path = get_database_path("apitest_listnet_ranker",
+				       db_index_two_documents);
+    string data_directory = test_driver::get_srcdir() + "/testdata/";
+    string query = data_directory + "query.txt";
+    string qrel = data_directory + "qrel.txt";
+    string training_data = data_directory + "training_data.txt";
+    Xapian::prepare_training_file(db_path, query, qrel, 10,
+				  "training_output.txt");
+    ifstream if1(training_data);
+    ifstream if2("training_output.txt");
+    string file1;
+    string file2;
+    while (getline(if1, file1)) {
+	 if (!getline(if2, file2)) {
+	     throw Xapian::InternalError("File Generated with incorrect "
+					 "inputs");
+	 }
+	istringstream iss1(file1);
+	istringstream iss2(file2);
+	string temp1;
+	string temp2;
+	int i = 0;
+	while ((iss1 >> temp1) && (iss2 >> temp2)) {
+	    if (i == 0 || i == 1 || i == 21) {
+		TEST_EQUAL(temp1, temp2);
+	    }
+	    else {
+		size_t t1 = temp1.find_first_of(':');
+		size_t t2 = temp2.find_first_of(':');
+		TEST(abs(stod(temp1.substr(t1 + 1)) -
+			 stod(temp2.substr(t2 + 1))) <
+		     0.001);
+	    }
+	    i++;
+	}
+	if (i < 22) {
+	    throw Xapian::InternalError("File Generated with incorrect "
+					"inputs");
```

Comment:
Better to use `FAIL_TEST("message");` to fail a test.

Or simpler use the `TEST` family of macros to perform the check:

```
TEST_REL(i, !=, 22);
```

(This will also report what value `i` actually had if the comparison fails).

And for the checks below:

```
TEST(!(iss2 >> temp2));
```

etc.